enhancements/installer/aws-lb-subnet-selection.md

```diff
@@ -0,0 +1,1008 @@
+---
+title: aws-lb-subnet-selection
+authors:
+- "@gcs278"
+reviewers:
+- "@candita"
+- "@frobware"
+- "@rfredette"
+- "@alebedev87"
+- "@miheer"
+- "@Miciah"
+- "@mtulio"
+approvers:
+- "@patrickdillon"
+api-approvers:
+- None
+creation-date: 2024-05-29
+last-updated: 2024-12-10
+tracking-link:
+  - https://issues.redhat.com/browse/CORS-3440
+see-also:
+  - "enhancements/ingress/lb-subnet-selection-aws.md"
+replaces:
+  - "enhancements/installer/aws-customer-provided-subnets.md"
+superseded-by:
+---
+
+# Installer LoadBalancer Subnet Selection for AWS
+
+## Summary
+
+This proposal enables the install-time configuration of subnets for the `default` IngressController
+as well as the default subnets for any user-created IngressController by extending the OpenShift
+Installer's install-config for AWS clusters. It also fixes a common issue where load balancers would
+map to unintended subnets ([OCPBUGS-17432](https://issues.redhat.com/browse/OCPBUGS-17432)).
+
+To achieve this, this enhancement deprecates the existing install-config field `platform.aws.subnets`
+in favor of a more flexible configuration field that handles specifying subnets for both
+IngressControllers and the cluster infrastructure. These new subnet roles, referred to as
+IngressController-role subnets and cluster-role subnets are further defined in [Defining Subnet Roles](#defining-subnets-roles).
+
+## Definitions and Terminology
+
+### The `default` IngressController
+
+This proposal will refer to the IngressController that gets created automatically during installation and handles
+the platform routes (console, auth, canary, etc.) as the `default` IngressController. Note that
+"default IngressController subnets" refers to the default subnets used by all IngressControllers, including
+the `default` IngressController. This is different from "`default`-IngressController subnets", which specifically
+means the subnets for the `default` IngressController.
+
+### User-Created IngressControllers
+
+A user-created IngressController is a custom IngressController that is created after installation (Day 2).
+It represents all IngressControllers except for the `default` IngressController.
+
+### Defining Subnets Roles
```

Comment:
I just heard from @makentenza that there is an RFE that would be positively impacted by the design of those fields: [RFE-4738](https://issues.redhat.com/browse/RFE-4738)

The request asks support to deploy LBs (control planes, ingress we'rent mentioned directly) on different subnets than EC2 instances/nodes, similar the initial request of this RFE.

If [those validations][1] goes well, I wonder if would be valid to expand/split the `Cluster` role value, which in currently design version covers API LB+Nodes, to something like this:
- ControlPlaneLB
- ControlPlaneNodes
- ComputeNodes

Or group the `*Nodes`, to something like `ClusterNodes`, so it may cover a use case of this RFE, where target deploy Control Plane LB into different subnets than nodes.

Example:

```yaml
        region: us-east-2
        subnetsConfig:
        - id: subnetID-premium-us-east-1a
          roles:
          - ControlPlaneLB
          - IngressController
        - id: subnetID-private-us-east-1a
          roles:
            - ControlPlaneNodes # or group "ClusterNodes"
            - ComputeNodes
```

LMK WDYT  @rvanderp3 @gcs278 @patrickdillon @makentenza 

[1]: https://github.com/openshift/enhancements/pull/1634#discussion_r1881057685

Reply:
@mtulio Good timing for this RFE while we are still talking through this design. 

I'm glad you reminded me about the ControlPlaneLBs. I spoke with service delivery, and the use case (dedicated LB subnets) also applies to the ControlPlaneLBs (_all_ LBs need to be in their own subnet). I completely overlooked the ControlPlaneLBs. The ControlPlaneLBs aren't affected by [OCPBUGS-17432](https://issues.redhat.com/browse/OCPBUGS-17432) interestingly enough due to the CAPI [logic](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/32ebcce58dbcc19cf83482443add80361aaf1ed2/pkg/cloud/services/elb/loadbalancer.go#L366) that only uses the cluster subnets. Service delivery confirmed there might be users with dedicated ControlPlaneLBs and dedicated IngressControllerLBs that are both different.

I'll admit, I don't quite understand that RFE's use case for IP space savings (routeable vs. non-routeable), but I think we need to this design anyways, so this proposal may provide what they need.

Implementing `ControlPlaneLB` looks straightforward since we already have a [field](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/63c66352579ef08bd365abd529e1924cde6f0d26/api/v1beta2/awscluster_types.go#L215) for it and I think similar validation from `IngressController` will apply to `ControlPlaneLB`. It's scope creep, but I think it's manageable and simple to add it.

As for splitting `ClusterNodes` into `ControlPlaneNodes` and `ComputeNodes`, I'm not sure if that's straight forward, as I don't totally understand how'd that be done with CAPI's [awsCluster.networkSpec.subnets](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/8f46a4d34d18cb090a4b48c9a9260cd3dffd9162/api/v1beta2/network_types.go#L340). Are there any RFEs or use cases to split compute & control plane nodes? I yield to the installer team for this design decision. I think we could always start with the `ClusterNodes` role and add the ControlPlane/Compute decomposed role in the future. That feels like an additive change that's compatible to me, but something we should consider.

Reply:
@mtulio I updated the proposal to include `ControlPlaneExternalLB`, `ControlPlaneInternalLB`, and `ClusterNode`. I think it was a good exercise that required me to learn a lot about how the installer and CAPA interact together, and how subnets are used. It also tested the API design, and I found a couple of issues that came from expanding the roles.

So, take a look and let me know what you think. I think I'm starting to arrive at the conclusion that it'd be much better to decompose the roles as best as we can on first pass of this enhancement, as adding new roles later may be tricky. As a result, this enhancement is growing. I also added added semantics of "automatic role selection", which is basically what `platform.aws.subnets` does today (it decides the purpose of the subnets). And I framed the roles as an "override" to that. Hopefully that helps frame the proposal more clearly.